File: golang/transformed-array/transformed-array.go
package main

import (
	"fmt"
	"reflect"
)

func constructTransformedArray(nums []int) []int {
	n := len(nums)
	result := make([]int, n)

	for i := 0; i < n; i++ {
		target_index := (i + nums[i]) % n

		if target_index < 0 {
			target_index += n
		}

		result[i] = nums[target_index]
	}

	return result
}

func runTest(testNum int, nums []int, expected []int) {
	result := constructTransformedArray(nums)

	status := "FAIL"
	if reflect.DeepEqual(result, expected) {
		status = "PASS"
	}

	fmt.Printf(
		"Test Case %d: Input %v -> Output %v (Expected: %v) -> %s\n",
		testNum, nums, result, expected, status,
	)
}

func main() {
	// Example 1
	runTest(1, []int{3, -2, 1, 1}, []int{1, 1, 1, 3})

	// Example 2
	runTest(2, []int{-1, 4, -1}, []int{-1, -1, 4})

	// Zero movement
	runTest(3, []int{0, 0, 0}, []int{0, 0, 0})

	// Full wrap around
	runTest(4, []int{3, 3, 3}, []int{3, 3, 3})
}

File: leetcode/transformed-array.py
class Solution(object):
    def constructTransformedArray(self, nums):

        n = len(nums)
        result= [0] * n

        for i in range(n):
            target_index = (i + nums[i]) % n
            result[i] = nums[target_index]
        
        return result
    
if __name__ == "__main__":
    solver = Solution()
    
    def run_test(case_num, nums, expected):
        result = solver.constructTransformedArray(nums)
        status = "PASS" if result == expected else "FAIL"
        print(f"Test Case {case_num}: Input {nums} -> Output {result} (Expected: {expected}) -> {status}")

    # Example 1
    # i=0: move 3 right -> index 3 (val 1)
    # i=1: move 2 left -> index 3 (val 1)
    # i=2: move 1 right -> index 3 (val 1)
    # i=3: move 1 right -> index 0 (val 3)
    run_test(1, [3,-2,1,1], [1,1,1,3])

    # Example 2
    # i=0: move 1 left -> index 2 (val -1)
    # i=1: move 4 right -> index 2 (val -1)
    # i=2: move 1 left -> index 1 (val 4)
    run_test(2, [-1,4,-1], [-1,-1,4])

    # Custom Case: Zero movement
    # [0, 0, 0] should return [0, 0, 0]
    run_test(3, [0, 0, 0], [0, 0, 0])

    # Custom Case: Full wrap around
    # length 3, move 3 steps right -> lands on same index
    run_test(4, [3, 3, 3], [3, 3, 3])